packages/library-base/src/commonMain/kotlin/io/realm/internal/VersionTracker.kt

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright 2022 Realm Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.realm.internal
+
+import io.realm.internal.interop.NativePointer
+import io.realm.internal.interop.RealmInterop
+import io.realm.internal.platform.WeakReference
+import kotlinx.atomicfu.AtomicRef
+import kotlinx.atomicfu.atomic
+
+/**
+ * Book keeping of intermediate versions that needs to be closed when no longer referenced or when
+ * explicitly closing a realm.
+ *
+ * NOTE: This is not thread safe, so synchronization should be enforced by the owner/caller.
+ */
+internal class VersionTracker(val log: RealmLog) {
+    // Set of currently open realms. Storing the native pointer explicitly to enable us to close
+    // the realm when the RealmReference is no longer referenced anymore.
+    private val intermediateReferences: AtomicRef<Set<Pair<NativePointer, WeakReference<RealmReference>>>> = atomic(mutableSetOf())
+
+    fun trackAndCloseExpiredReferences(realmReference: FrozenRealmReference? = null) {
+        val references = mutableSetOf<Pair<NativePointer, WeakReference<RealmReference>>>()
+        realmReference?.let {
+            references.add(Pair(realmReference.dbPointer, WeakReference(it)))
+        }
+        intermediateReferences.value.forEach { entry ->
+            val (pointer, ref) = entry
+            if (ref.get() == null) {
+                log.debug("Closing unreferenced version: ${RealmInterop.realm_get_version_id(pointer)}")
```

Comment:
We don't have access to anything sane info besides what we can retrieve from the raw native pointer. Only ending up here if the `RealmReference` is not referenced any longer. Holding on to it would keep it alive. But we could extend the map with some a link to the configuration if we really need it!?

Reply:
Good point. I'm not sure it is worth tracking extra things just for debugging, so let's just see if this is good enough.